From 682eb678991b3dde629df9178e7e088bbe4b7039 Mon Sep 17 00:00:00 2001
From: Frank Wunderlich <frank-w@public-files.de>
Date: Wed, 26 Mar 2025 19:44:18 +0100
Subject: [PATCH 36/84] net: ethernet: mtk_eth_soc: move desc assignment below
 check for its txd3

The field desc->txd3 should be checked for TX_DMA_OWNER_CPU flag before
assign a new desc.

old MTK-SDK uses same order:
https://git01.mediatek.com/plugins/gitiles/openwrt/feeds/mtk-openwrt-feeds/+/refs/heads/master/21.02/files/target/linux/mediatek/files-5.4/drivers/net/ethernet/mediatek/mtk_eth_soc.c#2895
new SDK seems not changing it.

Fixes: 656e705243fd ("net-next: mediatek: add support for MT7623 ethernet")
Signed-off-by: Frank Wunderlich <frank-w@public-files.de>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.c b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
index 9333f2b8e1d6..5b201c0214a0 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -2439,10 +2439,11 @@ static int mtk_poll_tx_qdma(struct mtk_eth *eth, int budget,
 	while ((cpu != dma) && budget) {
 		u32 next_cpu = desc->txd2;
 
-		desc = mtk_qdma_phys_to_virt(ring, desc->txd2);
 		if ((desc->txd3 & TX_DMA_OWNER_CPU) == 0)
 			break;
 
+		desc = mtk_qdma_phys_to_virt(ring, desc->txd2);
+
 		tx_buf = mtk_desc_to_tx_buf(ring, desc,
 					    eth->soc->tx.desc_size);
 		if (!tx_buf->data)
-- 
2.30.2

